process/dns_v2.go

```diff
@@ -0,0 +1,410 @@
+package process
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math"
+
+	"github.com/DataDog/mmh3"
+)
+
+// DNS data is encoded as a very basic bucketed hash table.  There are three blocks, or buffers, of data:
+//
+//	The "name" block is all of the unique DNS names.  The length of the name is stored as a varint
+//	followed by the name itself
+//
+//	The "bucket" block contains all of the hash buckets.  The format of each bucket is:
+//		varint for number of entries in bucket
+//		For each entry in the bucket:
+//			varint for length of ip
+//			ip bytes
+//			varint for number of names associated with the ip
+//			Each associated name is encoded as a varint which is the position of the actual name string in the name block
+//
+//	The "position" block is a list of varints, one for each bucket, where each varint is a pointer to the start
+//	of the bucket in the bucket block
+//
+// The overall buffer is encoded as:
+//	1 byte indicating version
+// 	2 bytes indicating the number of buckets
+//	varint indicating the length of the name buffer.
+//	varint indicating the length of the position buffer
+// 	varint indicating the position of the "middle" (bucketCount / 2) bucket in the position block
+//		We will use this to skip the half of the buckets when searching for the target bucket index
+//	position block
+//	bucket block
+//	name block
+//
+// Notes:
+//	Using varints saves space at the cost of not having random access to certain sections of data, particularly the
+//	bucket position mapping.  This was a deliberate trade off to reduce the size of the payload and thus memory usage
+//
+//	Varints are also more finicky to deal with in terms of calculating required space ahead of time.  This increases
+//	the implementation complexity, or at least the line count, but we reduce allocations & memory usage by
+// 	pre-sizing the output buffers
+//
+// This type is not thread safe
+
+type V2DNSEncoder struct {
+	BucketFactor float64
+	scratch      [binary.MaxVarintLen64]byte // Used for varint encoding
+}
+
+/*
+type bucketEntry struct {
+	keys []string
+	size int
+}
+*/
+// 1 byte for version, 2 byte for bucket count
+const dns1Version2PreambleLength = 3
+
+// Used for calculating the number of buckets for a given input map.
+// Currently the bucket count is calculated as `len(input) * bucketFactor`
+//const defaultBucketFactor = 0.75
+
+func NewV2DNSEncoder() DNSEncoder {
+	return &V2DNSEncoder{
+		BucketFactor: defaultBucketFactor,
+	}
+}
+func (e *V2DNSEncoder) Encode(dns map[string]*DNSEntry) ([]byte, error) {
+	return nil, fmt.Errorf("Encode not valid in V2")
+}
+
+func (e *V2DNSEncoder) EncodeMapped(dns map[string]*DNSDatabaseEntry) ([]byte, error) {
```

Comment:
Two questions:

- Earlier, we had this hash buckets for performance reasons as we were storing strings. Does it still make sense to have such buckets if we are only storing integers?
- Instead of storing indexes, can we directly store the starting position of the corresponding stringin the `EncodeDomainDatabase`? It will require more work during encoding but finding the right string during read would be much faster.

Reply:
We still do have hash buckets for decoding the mapping of IP address to DNSEntry.  We are still doing that mapping, although it is (as you point out) only for index. But it still makes the lookup faster.  Plus, my understanding is that there are places on the backend where this structure is relayed, and it's far more efficient to have it in a blob for that.

We could store starting positions, but IMO that complexity isn't worth it.  Right now you decode the array of strings once, and then index into it.  if we do what you say, each string access would involve creating a new string (rather than just pointing into the array).

Reply:
> Right now you decode the array of strings once, and then index into it.

My understanding is that we don't always need to decode that array of domains. In the `network-logger`, only domains that are part of the IP to domain mapping are relevant, not the ones associated with DNS stats. So decoding the whole array of strings is wasteful. Also, we have a backend limit of how many names we map per IP. So decoding names that go over that limit is unnecessary.
That is why a more efficient implementation of `IterateV2` that works similarly as the current `Iterate` would be helpful. If we can do that,  we won't have to make `network-logger` aware of the implementation details of how the encoding has changed. By using the starting position, the `IterateV2` function would be able to efficiently load only strings that are needed for an IP. Yes, that would additional complexity on the encoding part. But we encode once but decode a few times. So IMO that complexity is worth it.

@kevinconaway, since you are the one who implemented and is most familiar with the current encoding/decoding logic,  can you please share you take here and correct me if I am getting it all wrong?